pinot-core/src/main/java/org/apache/pinot/server/realtime/ControllerLeaderLocator.java

```diff
@@ -80,75 +87,127 @@ public static ControllerLeaderLocator getInstance() {
-          _controllerLeaderHostPort.getSecond());
-      return _controllerLeaderHostPort;
-    }
+    // No controller leader cached, fetches a fresh copy of external view and then gets the leader for the given table.
+    refreshControllerLeaderMap();
+    return _cachedControllerLeaderValid ? _cachedControllerLeaderMap.get(partitionId) : null;
   }
 
   /**
-   * Firstly checks whether lead controller resource has been enabled or not.
-   * If yes, use this as the leader for realtime segment completion once partition leader exists.
-   * Otherwise, try to use Helix leader.
-   * @param rawTableName table name without type.
-   * @return the controller leader id with hostname and port for this table, e.g. localhost_9000
+   * Checks whether lead controller resource has been enabled or not.
+   * If yes, updates lead controller pairs from the external view of lead controller resource.
+   * Otherwise, updates lead controller pairs from Helix cluster leader.
    */
-  private Pair<String, Integer> getLeaderForTable(String rawTableName) {
+  private void refreshControllerLeaderMap() {
     // Checks whether lead controller resource has been enabled or not.
     if (isLeadControllerResourceEnabled()) {
-      // Gets leader from lead controller resource.
-      return getLeaderFromLeadControllerResource(rawTableName);
+      refreshControllerLeaderMapFromLeadControllerResource();
     } else {
-      // Gets Helix leader to be the leader to this table, otherwise returns null.
-      return getHelixClusterLeader();
+      refreshControllerLeaderMapFromHelixClusterLeader();
     }
   }
 
   /**
-   * Checks whether lead controller resource is enabled or not. The switch is in resource config.
+   * Updates lead controller pairs from the external view of lead controller resource.
    */
-  private boolean isLeadControllerResourceEnabled() {
-    BaseDataAccessor<ZNRecord> dataAccessor = _helixManager.getHelixDataAccessor().getBaseDataAccessor();
-    Stat stat = new Stat();
+  private void refreshControllerLeaderMapFromLeadControllerResource() {
+    boolean refreshSucceeded = false;
     try {
-      ZNRecord znRecord = dataAccessor.get("/" + _helixManager.getClusterName() + "/CONFIGS/RESOURCE/"
-          + CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_NAME, stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
-      return Boolean.parseBoolean(znRecord.getSimpleField(CommonConstants.Helix.LEAD_CONTROLLER_RESOURCE_ENABLED_KEY));
+      ExternalView leadControllerResourceExternalView = _helixManager.getClusterManagmentTool()
+          .getResourceExternalView(_helixManager.getClusterName(), Helix.LEAD_CONTROLLER_RESOURCE_NAME);
+      if (leadControllerResourceExternalView == null) {
+        LOGGER.warn("External view of {} is null.", Helix.LEAD_CONTROLLER_RESOURCE_NAME);
+        return;
+      }
+      Set<String> partitionNames = leadControllerResourceExternalView.getPartitionSet();
+      if (partitionNames.isEmpty()) {
```

Comment:
(nit) This check can be merged into the next one

Reply:
Merged.